src/shared/openssl-util.c

```diff
@@ -1128,6 +1128,97 @@ int string_hashsum(
         return 0;
 }
 #  endif
+
+static int load_key_from_provider(
+                const char *provider,
+                const char *private_key_uri,
+                EVP_PKEY **ret_private_key) {
+
+        assert(provider);
+        assert(private_key_uri);
+        assert(ret_private_key);
+
+#if OPENSSL_VERSION_MAJOR >= 3
+        if (!OSSL_PROVIDER_try_load(/* ctx= */ NULL, provider, /* retain_fallbacks= */ true))
```

Comment:
You can use the store provider for all key type, no need to work with old APIs and then fallback

Reply:
providers are too new and not available everywhere, and also it's more difficult to make them work, so the fallback is still needed